usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -7302,62 +7327,67 @@ static CK_RV get_control_points_for_adapter(uint_32 adapter, uint_32 domain,
     size_t rlen, clen;
     long rc;
     CK_RV rv = 0;
-    ep11_target_t target;
+    target_t target;
     CK_IBM_XCP_INFO xcp_info;
     CK_ULONG xcp_info_len = sizeof(xcp_info);
 
-    memset(&target, 0, sizeof(target));
-    target.length = 1;
-    target.apqns[0] = adapter;
-    target.apqns[1] = domain;
+    rc = get_ep11_target_for_apqn(adapter, domain, &target);
+    if (rc != CKR_OK)
+        return rc;
 
     memset(cmd, 0, sizeof(cmd));
     rc = dll_xcpa_queryblock(cmd, sizeof(cmd), XCP_ADMQ_DOM_CTRLPOINTS,
                              (uint64_t) adapter << 32 | domain, NULL, 0);
     if (rc < 0) {
         TRACE_ERROR("%s xcpa_queryblock failed: rc=%ld\n", __func__, rc);
-        return CKR_DEVICE_ERROR;
+        rc = CKR_DEVICE_ERROR;
+        goto out;
     }
     clen = rc;
 
     memset(rsp, 0, sizeof(rsp));
     rlen = sizeof(rsp);
-    rc = dll_m_admin(rsp, &rlen, NULL, NULL, cmd, clen, NULL, 0,
-                     (uint64_t) & target);
+    rc = dll_m_admin(rsp, &rlen, NULL, NULL, cmd, clen, NULL, 0, target);
     if (rc < 0) {
```

Comment:
&target ???
A nice example why this target_t = uin64_t is in fact the wrong type.
The code now has to cast a pointer to the target list (uint64_t*) to uint64_t which is the target parameter type
and the compiler can't help any more.
It should be target_t = uint64_t * or even better const void *.

Reply:
O sorry - I got it. target is here in fact a target_t and so your line is fine.